File: go/day05/05.go
package day05

import (
	"slices"
	"strconv"
	"strings"
)

func parseSeedsLine(line string) []int {
	a := strings.Split(line, ": ")
	s := strings.Split(a[1], " ")

	seeds := []int{}
	for _, seed := range s {
		n, err := strconv.Atoi(seed)
		if err != nil {
			panic(err)
		}
		seeds = append(seeds, n)
	}

	return seeds
}

func parseMapNameLine(line string) [2]int {
	a := strings.Split(line, " ")
	b := strings.Split(a[0], "-to-")

	source, err := strconv.Atoi(b[0])
	if err != nil {
		panic(err)
	}

	destination, err := strconv.Atoi(b[1])
	if err != nil {
		panic(err)
	}

	return [...]int{source, destination}
}

func parseMapDataLine(line string) [3]int {
	a := strings.Split(line, " ")

	destination, err := strconv.Atoi(a[0])
	if err != nil {
		panic(err)
	}

	source, err := strconv.Atoi(a[1])
	if err != nil {
		panic(err)
	}

	length, err := strconv.Atoi(a[2])
	if err != nil {
		panic(err)
	}

	return [...]int{destination, source, length}
}

func findSeedLocation(seed int, dataMaps [][][3]int) int {
	curr := seed

	for _, dataMap := range dataMaps {
		for _, data := range dataMap {
			destination := data[0]
			source := data[1]
			length := data[2]

			if source <= curr && curr <= source+length {
				curr = destination + (curr - source)
				break
			}
		}
	}

	return curr
}

func Day05Puzzle1(lines []string) int {
	seeds := parseSeedsLine(lines[0])

	dataMaps := [][][3]int{}

	i := 2
	for i < len(lines) {
		dataMaps = append(dataMaps, [][3]int{})

		i += 1

		for i < len(lines) && len(lines[i]) != 0 {
			dataMaps[len(dataMaps)-1] = append(dataMaps[len(dataMaps)-1], parseMapDataLine(lines[i]))
			i += 1
		}

		i += 1
	}

	locations := []int{}
	for _, seed := range seeds {
		locations = append(locations, findSeedLocation(seed, dataMaps))
	}

	return slices.Min(locations)
}

File: python/05.py
from typing import List, Tuple

def parse_first_line(line: str):
    return list(map(int, line.rstrip().split(": ")[1].split()))

def process_data_maps(lines: List[str]):
    data_maps: List[List[Tuple[int, ...]]] = []
    for line in lines:
        line = line.rstrip()

        if len(line) == 0:
            data_maps.append([])
            continue

        if line.endswith("map:"):
            continue

        data_maps[-1].append(tuple(map(int, line.split())))

    return data_maps

def day_05_puzzle_1(inputs: List[str]):
    seeds = parse_first_line(inputs[0])
    data_maps = process_data_maps(inputs[1:])

    for data_map in data_maps:
        mapped = []
        for curr in seeds:
            for destination, source, length in data_map:
                if source <= curr <= source + length:
                    mapped.append(destination + curr - source)
                    break
            else:
                mapped.append(curr)
        seeds = mapped

    print(min(seeds))

def day_05_puzzle_2(inputs: List[str]):
    first_line = parse_first_line(inputs[0])
    data_maps = process_data_maps(inputs[1:])

    seeds: List[Tuple[int, int]] = []
    for i in range(0, len(first_line), 2):
        seeds.append((first_line[i],first_line[i] + first_line[i+1]))

    for data_map in data_maps:
        mapped: List[Tuple[int, int]] = []
        while len(seeds) > 0:
            seed_start, seed_end = seeds.pop()
            for destination_start, source_start, range_length in data_map:
                overlap_start = max(source_start, seed_start)
                overlap_end = min(source_start+range_length, seed_end)
                if overlap_start < overlap_end:
                    mapped.append((destination_start + overlap_start - source_start, destination_start + overlap_end - source_start))
                    if overlap_start > seed_start:
                        seeds.append((seed_start, overlap_start))
                    if overlap_end < seed_end: 
                        seeds.append((overlap_end, seed_end))
                    break
            else:
                mapped.append((seed_start, seed_end))
        seeds = mapped
    print(min(seeds))

if __name__ == "__main__":
    with open("../inputs/05.txt") as inputs:
        day_05_puzzle_2(inputs.readlines())
